### PR TITLE
Metaform statistics

### LINKED ISSUE #245: Metaform Statistics endpoint
As a user I'd like to be able to use a single endpoint for retrieving certain statistics about Metaforms and their Replies.

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #245

### CODE REVIEW

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt [hunk 1]
```diff
@@ -290,6 +290,23 @@ class ReplyDAO : AbstractDAO<Reply>() {
...
   }
 
+  /**
+   * Gets count of unprocessed replies by Metaform
+   *
+   * @param metaform metaform
+   * @return count of unprocessed replis
+   */
+  fun getCountOfUnprocessedReplies(metaform: Metaform): Long? {
+    return entityManager.createQuery(
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt:300]: Why this is written in plain SQL?

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt [hunk 2]
```diff
@@ -290,6 +290,23 @@ class ReplyDAO : AbstractDAO<Reply>() {
     return persist(reply)
   }
 
+  /**
+   * Gets count of unprocessed replies by Metaform
+   *
+   * @param metaform metaform
+   * @return count of unprocessed replis
+   */
+  fun getCountOfUnprocessedReplies(metaform: Metaform): Long? {
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt:299]:
```suggestion
  fun countUnprocessedReplies(metaform: Metaform): Long? {
```

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import java.time.OffsetDateTime
+import java.util.*
+import javax.persistence.Column
+import javax.persistence.Entity
+import javax.persistence.Id
+import javax.persistence.ManyToOne
+import javax.persistence.PrePersist
+
+@Entity
+class MetaformReplyCreated {
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt:12]: Add docs

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    @ManyToOne(optional = false)
+    lateinit var metaform: Metaform
+
+    @Column(nullable = false)
+    lateinit var replyCreated: OffsetDateTime
+
+    @ManyToOne(optional = false)
+    lateinit var reply: Reply
+
+    /**
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt:26]: It's a view, so it cannot have PrePersist handler

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import javax.persistence.Entity
+import javax.persistence.Id
+import javax.persistence.ManyToOne
+import javax.persistence.PrePersist
+
+@Entity
+class MetaformReplyCreated {
+
+    @Id
+    var id: UUID? = null
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt:15]: This field does not actually exist, so use reply as Id

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package fi.metatavu.metaform.server.persistence.model
+
+import java.time.OffsetDateTime
+import java.util.*
+import javax.persistence.*
+
+@Entity
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt:7]: Add docs

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+class MetaformReplyViewed {
+
+    @Id
+    var id: UUID? = null
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt:11]: This field does not actually exist, so use reply as Id

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    @Column(nullable = false)
+    lateinit var replyViewed: OffsetDateTime
+
+    @ManyToOne(optional = false)
+    lateinit var reply: Reply
+
+    /**
+     * JPA Pre-Persist handler
+     */
+    @PrePersist
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt:25]: It's a view, so it cannot have PrePersist handler

File: src/main/kotlin/fi/metatavu/metaform/server/rest/MetaformStatisticsApi.kt [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import fi.metatavu.metaform.server.controllers.MetaformStatisticsController
+import java.util.*
+import javax.enterprise.context.RequestScoped
+import javax.inject.Inject
+import javax.transaction.Transactional
+import javax.ws.rs.core.Response
+
+@RequestScoped
+@Transactional
+class MetaformStatisticsApi: fi.metatavu.metaform.api.spec.MetaformStatisticsApi, AbstractApi() {
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/metaform/server/rest/MetaformStatisticsApi.kt:13]: Add docs

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt (hunk 2 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt:299]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt (hunk 1 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/dao/ReplyDAO.kt:300]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt (hunk 3 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt:12]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt (hunk 4 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt:26]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt (hunk 5 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyCreated.kt:15]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt (hunk 6 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt:7]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt (hunk 7 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt:11]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt (hunk 8 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/persistence/model/MetaformReplyViewed.kt:25]: Resolved

File: src/main/kotlin/fi/metatavu/metaform/server/rest/MetaformStatisticsApi.kt (hunk 9 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/metaform/server/rest/MetaformStatisticsApi.kt:13]: Resolved

**Review by anttileppa — APPROVED** [review]

**Pull request merged**